### ISSUE TITLE
[New rule] Ensure dispose/close called

### ISSUE THREAD

**Issue by Ravensof** [dart-code-checker/dart-code-metrics#1229]
### Rule details

i want to be sure that my classes canceled/disposed, like Sink`s

### What type of rule is this?

Warns about a potential problem

### Example code

```dart
class CustomClass {
  void disposeFunction()
}

// bad

final custom = CustomClass() // lint, forgot to use disposeFunction

// good

final custom = CustomClass() // no lint
...
custom.disposeFunction()
```

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by incendial** [comment]
@Ravensof can you elaborate on when exactly dispose is supposed to be called? Right now it's not clear at all.

**Comment by Ravensof** [comment]
@incendial 
sorry, dont know how to say it correct.
i want it works something like lint rule for sinks https://dart-lang.github.io/linter/lints/close_sinks.html

just defines in analysis_options.yaml something like
```
dart_code_metrics:
  rules:
    close_disposable:
      -  DisposableMixin.dispose
```

then create class
```
class CustomClass with DisposableMixin { }
```

and when i create new object which is `DisposableMixin` there should be check that `dispose` ever called for that new object (or for variable where it was created)

**Comment by incendial** [comment]
Called where?

**Comment by Ravensof** [comment]
warning should be appear in places where disposable object were created

![image](https://user-images.githubusercontent.com/47619905/235864763-a4412e37-4c80-49ba-9424-d590c1aa41e2.png)

**Comment by incendial** [comment]
Got it, the fields. 
This rule is available in the Teams version https://dcm.dev/docs/teams/rules/flutter/dispose-fields/. No plans to add to the free version.

**Comment by Ravensof** [comment]
i dont know how its better in cases when disposable passed to another function.

in one side you have to dispose object in that level where you create it.
but dont know how to be here:
```
          Provider<Repository>(
            create: (context) => Repository(),
            dispose: (context, value) => value.dispose(),
          ),
```

maybe passing to another function should be counted as call `dispose` (like with lint close_sinks rule)

**Comment by incendial** [comment]
The case with providers is also covered by another Teams rule, that will become available in few days.

**Comment by Ravensof** [comment]
> Got it, the fields. This rule is available in the Teams version https://dcm.dev/docs/teams/rules/flutter/dispose-fields/. No plans to add to the free version.

thanks, that was fast, but how to apply it to my class? it just checking method named `dispose`?

**Comment by incendial** [comment]
Yes. After the release of 1.4.0, it will also check for 'close' and 'cancel'.

**Comment by Ravensof** [comment]
okay, thanks again

**Comment by incendial** [comment]
@Ravensof and the second part you've asked for https://dcm.dev/docs/teams/rules/provider/dispose-providers/, covers the `dispose` calls in Provider's dispose.

**Comment by Ravensof** [comment]
@incendial thanks, im appreciate it. it very useful, but we cant use teams version, i didnt even knew it exists -_-'

**Comment by incendial** [comment]
> but we cant use teams version,

What stops you from it?

**Comment by Ravensof** [comment]
> > but we cant use teams version,
> 
> What stops you from it?

because of politics visa/master_card not working in our country. 
even it works, for my application (which is hobby) its expensive, 
and for job i just cant say to my employeer that we need to buy teams version because of one rule :[

**Comment by incendial** [comment]
> and for job i just cant say to my employeer that we need to buy teams version because of one rule :[

Is it really only one rule?

**Comment by Ravensof** [comment]
i found some useful rules for myself, but im a regular programmer and i dont feel i have a chance to suggest it for use.
btw thanks for help 😅
